Shared heap

WenLY1 opened the pull request.

bianchui committed:
Enable merged os_mmap for aot data sections (#3681)

And enable merged os_mmap for aot data and text sections except on
platform nuttx and esp-idf.

Fix issue that aarch64 AOT module fails to load on android:
https://github.com/bytecodealliance/wasm-micro-runtime/issues/2274

WenLY1 committed:
implementation of shared heap

Signed-off-by: wenlingyun1 <wenlingyun1@xiaomi.com>

buddyliao commented on core/iwasm/libraries/shared-heap/shared_heap_wrapper.c:
```diff
@@ -0,0 +1,64 @@
+
+#define addr_native_to_app(ptr) \
+    wasm_runtime_addr_native_to_app(module_inst, ptr)
+
```
From line 7 ~ line 23, this code has already in wasm_export.h, can it be deleted here?

WenLY1 replied:
Done

wenyongh committed:
aot loader: Refine os_mmap related code (#3711)

wenyongh committed:
Merge pull request #3739 from bytecodealliance/main

Merge branch main into dev/merge_aot_data_text

wenyongh commented on build-scripts/config_common.cmake:
```diff
@@ -256,6 +256,14 @@ if (WAMR_BUILD_SHARED_MEMORY EQUAL 1)
+  message ("     Shared heap enabled")
+else ()
+  add_definitions (-DWASM_ENABLE_SHARED_HEAP=0)
+  message ("     Shared heap disabled")
```
最好把else分支去掉，不常用的feature一般不显示message，而core/config.h已经设置该宏为默认0，不需要再add_definitions

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -24,6 +24,10 @@ static Memory_Mode memory_mode = MEMORY_MODE_UNKNOWN;
 static mem_allocator_t pool_allocator = NULL;
 
+#if WASM_ENABLE_SHARED_HEAP != 0
+static WASMSharedHeap *shared_heap_list = NULL;
```
最好考虑有个lock来保护它？比如`static korp_mutex shared_heap_list_lock`，然后在wasm_runtime_memory_init里面初始化，在wasm_runtime_memory_destroy里面销毁。

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,6 +136,136 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+    WASMSharedHeap *heap =
+        runtime_malloc(heap_struct_size, error_buf, error_buf_size);
+
+    if (!(heap = runtime_malloc(heap_struct_size, error_buf, error_buf_size))) {
```
重复runtime_malloc，L169已经调用了一次

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,6 +136,136 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+        goto fail2;
+    }
+
+    size = align_uint(size, os_getpagesize());
```
这边要不要检查一下size，限制它的最大大小和最小大小

WenLY1 replied:
如何确认最大大小和最小大小，以什么为准呢

wenyongh replied:
可以参考一下app heap，感觉和它设置一样就可以吧:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/config.h#L390-L395

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,6 +136,136 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+    else {
+        heap->next = shared_heap_list;
+        shared_heap_list = heap;
+    }
```
将heap加到shared_heap_list最好加锁

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,6 +136,136 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+    return heap;
+
+fail4:
+    if (heap->base_addr) {
```
这边heap->base_addr肯定不为NULL，不需要这一句？

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,6 +136,136 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+    }
+
+fail3:
+    if (heap->heap_handle) {
```
same as above

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,6 +136,136 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+    WASMModuleInstance *wasm_module_inst = (WASMModuleInstance *)module_inst;
+    if (wasm_module_inst->e->shared_heap
+        && wasm_module_inst->e->shared_heap != heap) {
+        wasm_runtime_detach_shared_heap(module_inst);
```
instance已经attach了一个heap，感觉报错好点？让user自己先detach，再attach？

wenyongh replied:
另外有几个问题: (1) 整个attach或detach的流程要不要也加锁，(2) 对于多线程, 因为每个线程都有一个instance结构，但是它们共用线性内存，其中一个attach了shared heap，是不是也要同时attach shared heap到其它线程？ (3) 如果当前线程的线性内存的地址空间已经和shared heap的地址空间重叠了，应该禁止attach？(考虑LOG_WARNING然后返回失败)

另外wasm_memory_enlarge_internal里面，也应该检查一下，如果enlarge后的线性内存的地址空间和shared heap的地址空间重叠，感觉应该禁止enlarge (考虑LOG_WARNING然后返回失败)

WenLY1 replied:
关于问题2：有办法获取到其他线程的instance结构吗
关于问题3：如何判断线性内存空间和shared heap地址空间重叠呢

wenyongh replied:
问题2可以参考一下wasm_cluster_spread_custom_data和wasm_cluster_set_context:
https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/libraries/thread-mgr/thread_manager.c#L1373-L1445

问题3，以wasm32为例，当enlarge后的线性内存的大小大于 4G-shared_heap_size，就是重叠了，前者offset的最大范围是 0 到 4G-shared_heap_size-1, 后者是 4G-shared_heap_size 到 4G-1.

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,6 +136,136 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+        if (heap->alloc_type == Alloc_With_Pool) {
+            *p_native_addr = mem_allocator_malloc(
+                wasm_module_inst->e->shared_heap->heap_handle, size);
+            return addr_native_to_app(p_native_addr);
```
应该不能调用addr_native_to_app，shared heap的app offset范围是 [4G-shared_heap_size, 4G), (1) 如果*p_native_addr == NULL，返回0,  (2)否则应该是 `4G-shared_heap_size + (*p_native_addr - heap->base_addr)`
如果是memory64，考虑`UINT64_MAX-shared_heap_size+1 + (*p_native_addr - heap->base_addr)`?

WenLY1 replied:
addr_native_to_app这个函数中对shared heap做了处理，这样应该可以吧
![image](https://github.com/user-attachments/assets/889b5106-b32a-4118-86e4-a7355f9911b6)

wenyongh replied:
需要细看一下，我感觉是这边直接转换好点，那边还要做不少检查。

wenyongh replied:
最好直接
```C
    if (*p_native_addr == NULL)
        return 0;

#if WASM_ENABLE_MEMORY64 != 0
    if (memory->is_memory64)
        return UINT64_MAX - heap->size + 1 + ((uint8 *)*p_native_addr - heap->base_addr);
#endif

        return UINT32_MAX - heap->size + 1 + ((uint8 *)*p_native_addr - heap->base_addr);
```

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,6 +136,136 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+
+    if (wasm_module_inst->e->shared_heap) {
+        return mem_allocator_free(wasm_module_inst->e->shared_heap->base_addr,
+                                  addr_app_to_native(ptr));
```
同上，不能用addr_app_to_native

wenyongh replied:
最好:
```C
    WASMMemoryInstance *memory;
    memory = wasm_get_default_memory(wasm_module_inst);

#if WASM_ENABLE_MEMORY64 != 0
    if (memory->is_memory64) {
        if (ptr >= UINT64_MAX - heap->size)
            return heap->base_addr + (ptr - (UINT64_MAX - heap->size + 1));
        else
            return 0;
    }
#endif
      if (ptr >= UINT32_MAX - heap->size && ptr <= UINT32_MAX)
          return heap->base_addr + (ptr - (UINT32_MAX - heap->size + 1));
      else
          return 0;
```
```

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -174,6 +308,22 @@ wasm_runtime_memory_destroy()
+    WASMSharedHeap *heap = shared_heap_list;
+    WASMSharedHeap *cur;
+    while (heap) {
+        mem_allocator_destroy(heap->heap_handle);
```
感觉这句和下面几句放一起是不是好点？并且改成`mem_allocator_destroy(cur->heap_handle)`

wenyongh commented on core/iwasm/common/wasm_memory.h:
```diff
@@ -41,6 +41,36 @@ SET_LINEAR_MEMORY_SIZE(WASMMemoryInstance *memory, uint64 size)
+wasm_runtime_detach_shared_heap(WASMModuleInstanceCommon *module_inst);
+
+uint64
+wasm_runtime_shared_heap_malloc(wasm_module_inst_t module_inst, uint64 size,
+                                void **p_native_addr);
+
+void
+wasm_runtime_shared_heap_free(wasm_module_inst_t module_inst, uint64 ptr);
```
最好wasm_module_inst_t -> WASMModuleInstanceCommon *

wenyongh commented on core/iwasm/common/wasm_runtime_common.c:
```diff
@@ -173,6 +173,10 @@ static LLVMJITOptions llvm_jit_options = { 3, 3, 0, false };
 #endif
 
+#if WASM_ENABLE_SHARED_HEAP != 0
+static uint32 shared_heap_size = 0;
```
可能有多个shared heap，每个的大小都不一样，感觉这句以及下面的wasm_runtime_get_shared_heap_size等都不需要吧？由用户在每次wasm_runtime_create_shared_heap里面指定就好了

WenLY1 replied:
wasm_runtime_create_shared_heap这个接口会暴露给用户吗

wenyongh replied:
会的，参考文档:
```C
 - wasm_export.h
      struct WASMSharedHeap;
      typedef struct WASMSharedHeap wasm_shared_heap_t;
      export: wasm_runtime_create_shared_heap, wasm_runtime_destroy_shared_heap,
              wasm_rutnime_attach_shared_heap, wasm_runtime_detach_shared_heap
              wasm_runtime_shared_heap_malloc, wasm_runtime_shared_heap_free
```
比如先wasm_runtime_create_shared_heap, 然后把instance1, instance2, instance3 attach到它，
然后再wasm_runtime_create_shared_heap创建另一个shared heap, 然后把instance4, instance5, instance6 attach到它

WenLY1 replied:
> 会的，参考文档:
> 
> ```c
>  - wasm_export.h
>       struct WASMSharedHeap;
>       typedef struct WASMSharedHeap wasm_shared_heap_t;
>       export: wasm_runtime_create_shared_heap, wasm_runtime_destroy_shared_heap,
>               wasm_rutnime_attach_shared_heap, wasm_runtime_detach_shared_heap
>               wasm_runtime_shared_heap_malloc, wasm_runtime_shared_heap_free
> ```
> 
> 比如先wasm_runtime_create_shared_heap, 然后把instance1, instance2, instance3 attach到它， 然后再wasm_runtime_create_shared_heap创建另一个shared heap, 然后把instance4, instance5, instance6 attach到它

wasm_runtime_create_shared_heap/wasm_rutnime_attach_shared_heap/wasm_runtime_detach_shared_heap这几个API需要注册到wamr虚拟机吗，处理方式类似于shared_malloc/shared_free这两个接口吗

wenyongh replied:
这些API是要加到wasm_export.h，类似其它API

wenyongh commented on core/iwasm/include/wasm_export.h:
```diff
@@ -193,6 +193,22 @@ typedef enum RunningMode {
+    void *heap_handle;
+    uint8_t *base_addr;
+    uint32_t size;
+    mem_alloc_type_t alloc_type;
+    MemAllocOption alloc_option;
```
好像这两个field不需要用到

wenyongh commented on core/iwasm/include/wasm_export.h:
```diff
@@ -226,6 +242,9 @@ typedef struct RuntimeInitArgs {
     uint32_t segue_flags;
+
+    uint32_t shared_heap_size;
+    shared_heap_alloc_type_t shared_heap_alloc_type;
```
好像这个field不需要用到

wenyongh commented on core/iwasm/libraries/shared-heap/shared_heap_wrapper.c:
```diff
@@ -0,0 +1,50 @@
+#include "bh_common.h"
```
加一下license header

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,6 +136,136 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+    }
+
+    size = align_uint(size, os_getpagesize());
+    if (!(heap->base_addr = runtime_malloc(size, error_buf, error_buf_size))) {
```
现在线性内存默认是用os_mmap分配的，感觉shared heap是不是也这样好点？改成调用wasm_mmap_linear_memory？

yamt committed:
Add a comment on AOT_SECTION_TYPE_SIGNATURE (#3746)

cf. https://github.com/bytecodealliance/wasm-micro-runtime/issues/3744

Andersbakken committed:
Fix arm64 issues on mac (#3688)

Make wamrc normalize "arm64" to "aarch64v8". Previously the only way to
make the "arm64" target was to not specify a target on 64 bit arm-based
mac builds. Now arm64 and aarch64v8 are treated as the same.

Make aot_loader accept "aarch64v8" on arm-based apple (as well as
accepting legacy "arm64" based aot targets).

This also removes __APPLE__ and __MACH__ from the block that defaults
size_level to 1 since it doesn't seem to be supported for aarch64:
`LLVM ERROR: Only small, tiny and large code models are allowed on AArch64`

WenLY1 committed:
add lock to shared_heap

Signed-off-by: wenlingyun1 <wenlingyun1@xiaomi.com>

WenLY1 committed:
throw warning when a module is already attached to a shared heap

Signed-off-by: wenlingyun1 <wenlingyun1@xiaomi.com>

WenLY1 committed:
clang format adaption

Signed-off-by: wenlingyun1 <wenlingyun1@xiaomi.com>

WenLY1 committed:
add overlapping checks for linear memory and shared heap

Signed-off-by: wenlingyun1 <wenlingyun1@xiaomi.com>

no1wudi committed:
CI: Freeze version of bloaty for NuttX compilation (#3756)

Fix the compilation error of this CI:
https://github.com/bytecodealliance/wasm-micro-runtime/actions/runs/10575515238

```
/__w/wasm-micro-runtime/wasm-micro-runtime/bloaty/third_party/abseil-cpp/absl/debugging/failure_signal_handler.cc:139:32: error: no matching function for call to 'max(long int, int)'
  139 |   size_t stack_size = (std::max(SIGSTKSZ, 65536) + page_mask) & ~page_mask;
      |                        ~~~~~~~~^~~~~~~~~~~~~~~~~
```

wenyongh committed:
Merge branch 'main' into dev/merge_aot_data_text

wenyongh committed:
Merge pull request #3757 from bytecodealliance/dev/merge_aot_data_text

Merge branch dev/merge_aot_data_text into main to keep the commit history.

TianlongLiang committed:
aot compiler: Allow to control stack boundary check when boundary check is enabled (#3754)

In the AOT compiler, allow the user to control stack boundary check when the boundary
check is enabled (e.g. `wamrc --bounds-checks=1`). Now the code logic is:

1. When `--stack-bounds-checks` is not set, it will be the same value as `--bounds-checks`.
2. When `--stack-bounds-checks` is set, it will be the option value no matter what the
    status of `--bounds-checks` is.

Andersbakken committed:
aot loader: Call os_mmap with MMAP_MAP_32BIT only when target is x86-64 or riscv64 (#3755)

Mac on aarch64 uses posix_memmap.c os_mmap which doesn't do anything with
the flag MMAP_MAP_32BIT for that build so this condition ends up asserting unless
the mapping ends up in the first 4 gigs worth of addressable space.

Thsi PR changes to call os_mmap with MMAP_MAP_32BIT flag only when the target
is x86-64 or riscv64, and the macro __APPLE__ isn't enabled. The behavior is similar
to what the posix os_mmap does.

loganek committed:
Update ref to the multi-memory tests (#3764)

The specific commit has been deleted, I am pointing to the same commit in the main branch though.

sjamesr committed:
Improve posix mmap retry logic (#3714)

- Only retry on EAGAIN, ENOMEM or EINTR.
- On EINTR, don't count it against the retry budget, just keep retrying.
  EINTR can happen in bursts.
- Log the errno on failure, and don't conditionalize that logging on
  BH_ENABLE_TRACE_MMAP. In other parts of the code, error logging is not
  conditional on that define, while turning on that tracing define makes
  things overly verbose.

no1wudi committed:
compilation_on_nuttx.yml: Update checkout action to suppress warnings (#3765)

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,10 +137,174 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+    }
+
+    if (shared_heap_list == NULL) {
+        os_mutex_lock(&shared_heap_list_lock);
```
建议把os_mutex_lock/unlock提到`if () {..} else {..}`的外面，减少点代码？

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,10 +137,174 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+{
+    WASMModuleInstance *wasm_module_inst = (WASMModuleInstance *)module_inst;
+    uint32 linear_mem_size = 0;
+    WASMMemoryInstance *memory = NULL;
```
这行后面最好加一行空格，代码风格好点。

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,10 +137,174 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+    uint32 linear_mem_size = 0;
+    WASMMemoryInstance *memory = NULL;
+    if (wasm_module_inst->e->shared_heap
+        && wasm_module_inst->e->shared_heap != heap) {
```
如果`wasm_module_inst->e->shared_heap == heap`，要吗直接返回false，要吗直接返回true, 不需要再进行后面操作。似乎返回true合理点？这边代码改成:
```C
    if (wasm_module_inst->e->shared_heap) {
        if (wasm_module_inst->e->shared_heap != heap) {
            LOG_WARNING("A shared heap is already attached");
            return false;
        }
        return true;
    }
```

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,10 +137,174 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+    // check if linear memory and shared heap are overlapped
+
+    memory = wasm_get_default_memory(wasm_module_inst);
+    linear_mem_size += memory->num_bytes_per_page * memory->cur_page_count;
```
`linear_mem_size +=`改成`linear_mem_size =`好点，否则代码怪怪的。
另外可以 `linear_mem_size = memory->memory_data_size;`

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,10 +137,174 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+                                WASMSharedHeap *heap)
+{
+    WASMModuleInstance *wasm_module_inst = (WASMModuleInstance *)module_inst;
+    uint32 linear_mem_size = 0;
```
uint32 -> uint64，即使memory32，大小也可能是4G = UINT32_MAX + 1，需要用64位表示

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,10 +137,174 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+        LOG_WARNING("Linear memory address is overlapped with shared heap");
+        return false;
+    }
+#endif
```
使用`memory->is_memory64`来判断，而不是直接使用宏，我理解可以类似下面代码:
```C
    if (false
#if WASM_ENABLE_MEMORY64 != 0
        || (memory->is_memory64 && linear_mem_size > UINT64_MAX - heap->size)
#endif
        || (!memory->is_memory64 && linear_mem_size > UINT32_MAX - heap->size)
    } {
        ...
    }
```

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,10 +137,174 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+    if (heap) {
+        if (heap->alloc_type == Alloc_With_Pool) {
+            *p_native_addr = mem_allocator_malloc(
+                wasm_module_inst->e->shared_heap->heap_handle, size);
```
可以使用`heap->heap_handle`

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,10 +137,174 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+    WASMModuleInstance *wasm_module_inst = (WASMModuleInstance *)module_inst;
+    WASMSharedHeap *heap = wasm_module_inst->e->shared_heap;
+    if (heap) {
+        if (heap->alloc_type == Alloc_With_Pool) {
```
现在alloc_type是不是没有用的，因为都是os_mmap分配的？可以去掉这个判断？

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,10 +137,174 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+    }
+#endif
     if (mem_alloc_type == Alloc_With_Pool) {
         return wasm_memory_init_with_pool(alloc_option->pool.heap_buf,
```
return false之前应该os_mutex_destroy(&shared_heap_list_lock)，最好:
```C
    bool ret;

#if WASM_ENABLE_SHARED_HEAP != 0
    if (os_mutex_init(&shared_heap_list_lock)) {
        return false;
    }
#endif

    if (mem_alloc_type == Alloc_With_Pool) {
        ret = wasm_memory_init_with_pool(...);
    }
    else if (mem_alloc_type == Alloc_With_Allocator) {
        ret = wasm_memory_init_with_allocator(...);
    }
    else if (mem_alloc_type == Alloc_With_System_Allocator) {
        memory_mode = MEMORY_MODE_SYSTEM_ALLOCATOR;
        return true;
    }
    else {
        ret = false;
    }

    if (!ret)
        os_mutex_destroy(&shared_heap_list_lock)

    return ret;
```

wenyongh commented on core/iwasm/interpreter/wasm_runtime.c:
```diff
@@ -3257,6 +3270,10 @@ wasm_deinstantiate(WASMModuleInstance *module_inst, bool is_sub_inst)
 #endif
 
+#if WASM_ENABLE_SHARED_HEAP != 0
+    wasm_runtime_shared_heap_destroy();
```
不是在这里调用这个函数，应该在wasm_runtime_memory_destroy里面destroy所有的shared heap.

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -174,6 +343,22 @@ wasm_runtime_memory_destroy()
 }
+#if WASM_ENABLE_SHARED_HEAP != 0
+void
+wasm_runtime_shared_heap_destroy()
```
最好改成`static void shared_heap_destroy()`，然后在wasm_runtime_memory_destroy里面调用它

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -174,6 +343,22 @@ wasm_runtime_memory_destroy()
+    while (heap) {
+        cur = heap;
+        heap = heap->next;
+        mem_allocator_destroy(heap->heap_handle);
```
应该是`cur->heap_handle`

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -315,6 +500,48 @@ wasm_runtime_get_mem_alloc_info(mem_alloc_info_t *mem_alloc_info)
+    WASMSharedHeap *heap = module_inst->e->shared_heap;
+
+    if (!is_memory64) {
+        if (app_offset >= UINT32_MAX - heap->size
```
shared heap的offset范围是`4G - shared_heap_size`到`4G - 1` (=UINT32_MAX)，而不是`UINT32_MAX - shared_heap_size`到`4G`。这边应该是` >= UINT32_MAX - heap->size + 1`

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -315,6 +500,48 @@ wasm_runtime_get_mem_alloc_info(mem_alloc_info_t *mem_alloc_info)
+        }
+    }
+    else {
+        if (app_offset >= UINT64_MAX - heap->size
```
同上

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -315,6 +500,48 @@ wasm_runtime_get_mem_alloc_info(mem_alloc_info_t *mem_alloc_info)
+    WASMSharedHeap *heap = module_inst->e->shared_heap;
+
+    if (heap && heap->base_addr && addr + bytes <= heap->base_addr + heap->size
+        && addr + bytes >= heap->base_addr) {
```
addr + bytes可能integer overflow。似乎:
```C
addr >= heap->base_addr && addr + bytes <= heap->base_addr + heap->size && addr + bytes > addr
```
好些？

wenyongh commented on core/iwasm/common/wasm_memory.h:
```diff
@@ -41,6 +41,35 @@ SET_LINEAR_MEMORY_SIZE(WASMMemoryInstance *memory, uint64 size)
+                              uint64 ptr);
+
+void
+wasm_runtime_shared_heap_destroy();
```
去掉这个，不需要export它

wenyongh commented on core/iwasm/interpreter/wasm_runtime.c:
```diff
@@ -2289,6 +2289,16 @@ wasm_instantiate(WASMModule *module, WASMModuleInstance *parent,
+    WASMSharedHeap *heap =
+        wasm_runtime_create_shared_heap(&args, error_buf, error_buf_size);
+    wasm_runtime_attach_shared_heap((WASMModuleInstanceCommon *)module_inst,
+                                    heap);
```
不需要在这里create并attach shared heap，去掉这部分代码

wenyongh commented on core/iwasm/interpreter/wasm_runtime.c:
```diff
@@ -3735,6 +3811,22 @@ wasm_module_dup_data(WASMModuleInstance *module_inst, const char *src,
+                          void **p_native_addr)
+{
+    return wasm_module_shared_malloc_internal(module_inst, NULL, size,
+                                              p_native_addr);
```
wasm_memory.c已经定义了wasm_runtime_shared_heap_malloc了，按理调用它就可以了

wenyongh commented on core/iwasm/interpreter/wasm_runtime.c:
```diff
@@ -3735,6 +3811,22 @@ wasm_module_dup_data(WASMModuleInstance *module_inst, const char *src,
+void
+wasm_module_shared_free(WASMModuleInstance *module_inst, uint64 ptr)
+{
+    wasm_module_shared_free_internal(module_inst, NULL, ptr);
```
同上

loganek committed:
Remove unnecessary code duplication in aot runtime (#3767)

WenLY1 committed:
attach shared heap to thread cluster and fix some issue

Signed-off-by: wenlingyun1 <wenlingyun1@xiaomi.com>

loganek committed:
refactoring: Re-use commit IP functionality between exception handling and other cases (#3768)

loganek committed:
Add wamrc parameter to configure stack frame features (#3763)

Those parameters can be used to reduce the size of the AOT code.

There's going to be more changes related to AOT code size reduction,
this is just the initial step.

p.s. https://github.com/bytecodealliance/wasm-micro-runtime/issues/3758

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,16 +141,206 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+{
+    if (error_buf != NULL) {
+        snprintf(error_buf, error_buf_size,
+                 "WASM module instantiate failed: %s", string);
```
这个不是在instantiate时报的错，看起来只被create shared heap调用，感觉换个字符串合适点？

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,16 +141,206 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+
+    // check if linear memory and shared heap are overlapped
+    memory = wasm_get_default_memory(wasm_module_inst);
+    linear_mem_size = linear_mem_size + memory->memory_data_size;
```
这样写有点奇怪，建议就 `linear_mem_size = memory->memory_data_size;`

WenLY1 replied:
需要考虑multi-memory的情况吗，这种情况下应该将所有memory的size累加吗

wenyongh replied:
不需要考虑multi-memory，也不能把所有的memory的size累加，先只针对default memory就可以了。

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,16 +141,206 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+        if (*p_native_addr == NULL)
+            return 0;
+#if WASM_ENABLE_MEMORY64 != 0
+        if (memory->is_memory64)
```
memory未定义，需要`WASMMemoryInstance *memory;  memory = wasm_get_default_memory(wasm_module_inst);`

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -475,6 +763,27 @@ wasm_runtime_addr_app_to_native(WASMModuleInstanceCommon *module_inst_comm,
+#if WASM_ENABLE_SHARED_HEAP != 0
+#if WASM_ENABLE_MEMORY64 == 0
+        else if (app_offset >= UINT32_MAX - module_inst->e->shared_heap->size
+                 && app_offset <= UINT32_MAX - 1) {
```
要加一下 `!memory_inst->is_memory64`，另外条件似乎不对，改成:
```C
        else if (!memory_inst->is_memory64
                 && app_offset >= UINT32_MAX - module_inst->e->shared_heap->size + 1
                 && app_offset <= UINT32_MAX)
```

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -475,6 +763,27 @@ wasm_runtime_addr_app_to_native(WASMModuleInstanceCommon *module_inst_comm,
+        else if (app_offset >= UINT32_MAX - module_inst->e->shared_heap->size
+                 && app_offset <= UINT32_MAX - 1) {
+            uint64 heap_start = (uint64)(UINT32_MAX)
+                                - (uint64)module_inst->e->shared_heap->size;
```
`(uint64)(UINT32_MAX) - (uint64)module_inst->e->shared_heap->size + 1`，缺少`+ 1`

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -475,6 +763,27 @@ wasm_runtime_addr_app_to_native(WASMModuleInstanceCommon *module_inst_comm,
+        }
+#else
+        else if (app_offset >= UINT64_MAX - module_inst->e->shared_heap->size
+                 && app_offset <= UINT64_MAX - 1) {
```
同上
```C
        else if (memory_inst->is_memory64
                 && app_offset >= UINT64_MAX - module_inst->e->shared_heap->size + 1
                 && app_offset <= UINT64_MAX)
```

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -475,6 +763,27 @@ wasm_runtime_addr_app_to_native(WASMModuleInstanceCommon *module_inst_comm,
+        else if (app_offset >= UINT64_MAX - module_inst->e->shared_heap->size
+                 && app_offset <= UINT64_MAX - 1) {
+            uint64 heap_start =
+                UINT64_MAX - (uint64)module_inst->e->shared_heap->size;
```
`UINT64_MAX - (uint64)module_inst->e->shared_heap->size + 1`

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -475,6 +763,27 @@ wasm_runtime_addr_app_to_native(WASMModuleInstanceCommon *module_inst_comm,
+                                - (uint64)module_inst->e->shared_heap->size;
+            uint64 heap_offset = (uint64)app_offset - heap_start;
+            addr = module_inst->e->shared_heap->base_addr + heap_offset;
+            return addr;
```
返回前先SHARED_MEMORY_UNLOCK(memory_inst);

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -475,6 +763,27 @@ wasm_runtime_addr_app_to_native(WASMModuleInstanceCommon *module_inst_comm,
+                UINT64_MAX - (uint64)module_inst->e->shared_heap->size;
+            uint64 heap_offset = (uint64)app_offset - heap_start;
+            addr = module_inst->e->shared_heap->data + heap_offset;
+            return addr;
```
返回前先SHARED_MEMORY_UNLOCK(memory_inst);

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -513,6 +827,23 @@ wasm_runtime_addr_native_to_app(WASMModuleInstanceCommon *module_inst_comm,
+            uint64 shared_heap_start = 0;
+#if WASM_ENABLE_MEMORY64 == 0
+            shared_heap_start =
+                UINT32_MAX - (uint64)module_inst->e->shared_heap->size;
```
缺了`+ 1`，并且要判断`!memory_inst->is_memory64`

WenLY1 replied:
为shared heap的native addr to app和 app offset to native addr封装一些函数会不会好一些

wenyongh replied:
嗯，那样更好，不容易出错，注意是专门给shared heap封装的，另外不只是app offset和native addr的相互转换，可能还包括获取起始offset ( = UINT32_MAX/UINT64_MAX - shared_heap_size +1)。你改改试试吧？

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -513,6 +827,23 @@ wasm_runtime_addr_native_to_app(WASMModuleInstanceCommon *module_inst_comm,
+                UINT32_MAX - (uint64)module_inst->e->shared_heap->size;
+#else
+            shared_heap_start =
+                UINT64_MAX - (uint64)module_inst->e->shared_heap->size;
```
缺了`+ 1`，并且要判断`memory_inst->is_memory64`

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -797,6 +1133,18 @@ wasm_enlarge_memory_internal(WASMModuleInstance *module, uint32 inc_page_count,
+    heap = module->e->shared_heap;
+#if WASM_ENABLE_MEMORY64 != 0
+    if (total_size_new > UINT64_MAX - heap->size) {
+        LOG_WARNING("Linear memory address is overlapped with shared heap");
```
要返回失败，另外要判断`memory_inst->is_memory64`

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -797,6 +1133,18 @@ wasm_enlarge_memory_internal(WASMModuleInstance *module, uint32 inc_page_count,
+    }
+#else
+    if (total_size_new > UINT32_MAX - heap->size) {
+        LOG_WARNING("Linear memory address is overlapped with shared heap");
```
要返回失败，另外要判断`!memory_inst->is_memory64`

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,16 +141,206 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+wasm_runtime_shared_heap_malloc(wasm_module_inst_t module_inst, uint64_t size,
+                                void **p_native_addr)
+{
+    WASMModuleInstance *wasm_module_inst = (WASMModuleInstance *)module_inst;
```
这个假定了module_inst是interpreter模式的，它可能是AOT的，最好区分一下，AOT可以先考虑不弄

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,16 +141,206 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+{
+    WASMModuleInstance *wasm_module_inst = (WASMModuleInstance *)module_inst;
+
+    if (wasm_module_inst->e->shared_heap) {
```
同上这个假定了module_inst是interpreter模式的，它可能是AOT的，最好区分一下，AOT可以先考虑不弄

wenyongh commented on core/iwasm/interpreter/wasm_runtime.c:
```diff
@@ -3735,6 +3738,22 @@ wasm_module_dup_data(WASMModuleInstance *module_inst, const char *src,
+#if WASM_ENABLE_SHARED_HEAP != 0
+uint64
+wasm_module_shared_malloc(WASMModuleInstance *module_inst, uint64 size,
+                          void **p_native_addr)
```
有wasm_runtime_shared_heap_malloc就可以了，我理解这边wasm_module_shared_malloc这个函数可以删掉不要？

wenyongh commented on core/iwasm/common/wasm_runtime_common.c:
```diff
@@ -3289,6 +3289,38 @@ wasm_runtime_module_dup_data(WASMModuleInstanceCommon *module_inst,
 
+#if WASM_ENABLE_SHARED_HEAP != 0
+uint64
+wasm_runtime_module_shared_malloc(WASMModuleInstanceCommon *module_inst,
```
有wasm_runtime_shared_heap_malloc就可以了，这边wasm_runtime_module_shared_malloc/free可以删掉不要？

wenyongh commented on core/iwasm/libraries/shared-heap/shared_heap_wrapper.c:
```diff
@@ -0,0 +1,55 @@
+    wasm_runtime_validate_native_addr(module_inst, addr, size)
+
+#define module_shared_malloc(size, p_native_addr) \
+    wasm_runtime_module_shared_malloc(module_inst, size, p_native_addr)
```
最好用wasm_runtime_shared_heap_malloc

wenyongh commented on core/iwasm/libraries/shared-heap/shared_heap_wrapper.c:
```diff
@@ -0,0 +1,55 @@
+    wasm_runtime_module_shared_malloc(module_inst, size, p_native_addr)
+
+#define module_shared_free(offset) \
+    wasm_runtime_module_shared_free(module_inst, offset)
```
最好调用wasm_runtime_shared_heap_free

mattyg committed:
Fix building iwasm_shared and iwasm_static libs on win32 (#3762)

Fixes to enable building iwasm_shared and iwasm_static libraries on win32.

WenLY1 committed:
fix some issues

Signed-off-by: wenlingyun1 <wenlingyun1@xiaomi.com>

WenLY1 committed:
add function for converting native addr to app offset

Signed-off-by: wenlingyun1 <wenlingyun1@xiaomi.com>

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,16 +141,360 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+    }
+
+    size = align_uint(size, os_getpagesize());
+    if (size > UINT32_MAX - APP_HEAP_SIZE_MAX + 1 || size < APP_HEAP_SIZE_MIN) {
```
`size > UINT32_MAX - APP_HEAP_SIZE_MAX + 1` 不合理，应该是`if (size > APP_HEAP_SIZE_MAX || size < APP_HEAP_SIZE_MIN)`

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,16 +141,360 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+    size = align_uint(size, os_getpagesize());
+    if (size > UINT32_MAX - APP_HEAP_SIZE_MAX + 1 || size < APP_HEAP_SIZE_MIN) {
+        set_error_buf(error_buf, error_buf_size, "invalid size of shared heap");
+        goto fail1;
```
应该是goto fail3

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,16 +141,360 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+              || module_inst_comm->module_type == Wasm_Module_AoT);
+    WASMSharedHeap *heap = module_inst->e->shared_heap;
+
+    if (heap && heap->base_addr && addr >= heap->base_addr
```
似乎可以把`&& heap->base_addr`去掉，因为一旦heap被创建，肯定base_addr是被mmap出来了

wenyongh commented on core/iwasm/include/wasm_export.h:
```diff
@@ -193,6 +193,20 @@ typedef enum RunningMode {
+typedef struct WASMSharedHeap {
+    void *heap_handle;
+    uint8_t *base_addr;
+    uint32_t size;
```
我在想，`UINT32_MAX - heap->size + 1`和`UINT64_MAX - heap->size + 1`被经常用到，要不我们这边存一下它？比如`uint64 start_offset`，然后创建的时候，根据default memory的is_memory64的值，给它赋值，然后后面就可以拿出来用了:
```C
    if (memory->is_memory64)
        start_offset = UINT64_MAX - heap->size + 1;
    else
        start_offset = UINT32_MAX - heap->size + 1;
```

WenLY1 replied:
存在一个static 全局变量中吗

wenyongh replied:
不是，就存在WASMSharedHeap 结构里面，wasm_runtime_create_shared_heap的时候赋一下值

wenyongh replied:
另外这个结构不需要暴露在wasm_export.h里面，按理在wasm_memory.h就可以了？

WenLY1 replied:
> 另外这个结构不需要暴露在wasm_export.h里面，按理在wasm_memory.h就可以了？

这个比较麻烦，刚刚试了下，如果把WasmSharedHeap定义放到wasm_memory.h，头文件会循环依赖，wasm_runtime.h->wasm_memory.h->wasm_shared_memory.h->wasm_runtime.h。可以将WasmSharedHeap定义放到wasm_runtime_common.h中吗

wenyongh replied:
直接放到wasm_runtime.h呢？

WenLY1 replied:
> 直接放到wasm_runtime.h呢？

这样也可以，不过aot_runtime.h也需要一份WasmSharedHeap的定义

WenLY1 replied:
应该可以，我看到aot_runtime.h include了wasm_runtime.h

WenLY1 committed:
fix some issues

Signed-off-by: wenlingyun1 <wenlingyun1@xiaomi.com>

no1wudi committed:
CI: Disable parallel test in spectest for NuttX (#3780)

lum1n0us committed:
wasi-nn: Add a new target for llama.cpp as a wasi-nn backend (#3709)

Minimum support:
- [x] accept (WasmEdge) customized model parameters. metadata.
- [x] Target [wasmedge-ggml examples](https://github.com/second-state/WasmEdge-WASINN-examples/tree/master/wasmedge-ggml)
  - [x] basic
  - [x] chatml
  - [x] gemma
  - [x] llama
  - [x] qwen

---

In the future, to support if required:
- [ ] Target [wasmedge-ggml examples](https://github.com/second-state/WasmEdge-WASINN-examples/tree/master/wasmedge-ggml)
  - [ ] command-r. (>70G memory requirement)
  - [ ] embedding. (embedding mode)
  - [ ] grammar. (use the grammar option to constrain the model to generate the JSON output)
  - [ ] llama-stream. (new APIS `compute_single`, `get_output_single`, `fini_single`)
  - [ ] llava. (image representation)
  - [ ] llava-base64-stream. (image representation)
  - [ ] multimodel. (image representation)
- [ ] Target [llamaedge](https://github.com/LlamaEdge/LlamaEdge)

loganek committed:
AOT call stack optimizations (#3773)

- Implement TINY / STANDARD frame modes - tiny mode is only able to keep track on the IP
  and func idx, STANDARD mode provides more capabilities (parameters, stack pointer etc.).
- Implement FRAME_PER_FUNCTION / FRAME_PER_CALL modes - frame per function adds
  code at the beginning and at the end of each function for allocating / deallocating stack frame,
  whereas in per-call mode the frame is allocated before each call. The exception is call to
  the imported function, where frame-per-function mode also allocates the stack before the
  `call` instruction (as it can't instrument the imported function).

At the moment TINY + FRAME_PER_FUNCTION is automatically enabled in case GC and perf
profiling are disabled and `values` call stack feature is not requested. In all the other cases
STANDARD + FRAME_PER_CALL is used.

STANDARD + FRAME_PER_FUNCTION and TINY + FRAME_PER_CALL are currently not
implemented but possible, and might be enabled in the future.

ps. https://github.com/bytecodealliance/wasm-micro-runtime/issues/3758

yamt committed:
Appease GCC strict prototypes warning (#3775)

bnason-nf committed:
Fix compile error when multi-module and tags are enabled (#3781)

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,16 +141,364 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+    if (module_inst->module_type == Wasm_Module_Bytecode) {
+        if (((WASMModuleInstance *)module_inst)->e->shared_heap) {
+            LOG_WARNING("A shared heap is already attached");
+            return true;
```
should be `return false`

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,16 +141,364 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+}
+
+bool
+wasm_runtime_detach_shared_heap(WASMModuleInstanceCommon *module_inst)
```
感觉wasm_runtime_detach_shared_heap, wasm_runtime_detach_shared_heap_internal和wasm_cluster_detach_shared_heap都做成返回值就可以，如果下面的wasm_runtime_detach_shared_heap_internal只是把module_inst的shared heap句柄设成NULL的话。
否则这边都是直接return true，不大需要。

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,16 +141,364 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+{
+    WASMModuleInstance *module_inst = (WASMModuleInstance *)module_inst_comm;
+    bh_assert(module_inst_comm->module_type == Wasm_Module_Bytecode
+              || module_inst_comm->module_type == Wasm_Module_AoT);
```
如果上面is_app_addr_in_shared_heap没有bh_assert，这边也不assert吧，并且取heap的代码和上面最好一致(区分interpreter module inst和aot module inst)

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,16 +141,364 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+        return NULL;
+    }
+
+    if (is_app_addr_in_shared_heap(module_inst, memory->is_memory64, ptr, 1)) {
```
上面的shared_heap_addr_native_to_app没有检查native addr有没有在shared heap里面，这边最好也去掉？

wenyongh commented on core/iwasm/common/wasm_memory.h:
```diff
@@ -41,6 +41,36 @@ SET_LINEAR_MEMORY_SIZE(WASMMemoryInstance *memory, uint64 size)
+#if WASM_ENABLE_SHARED_HEAP != 0
+typedef struct SharedHeapInitArgs {
+    uint32 size;
+} SharedHeapInitArgs;
```
这个定义需要放在wasm_export.h，这边不保留
另外几个函数也要在wasm_export.h声明出来（参考其它函数），这个文件的声明也保留:

wasm_runtime_create_shared_heap
wasm_runtime_attach_shared_heap
wasm_runtime_detach_shared_heap
wasm_runtime_shared_heap_malloc
wasm_runtime_shared_heap_free

wenyongh commented on core/iwasm/common/wasm_runtime_common.c:
```diff
@@ -3289,6 +3289,38 @@ wasm_runtime_module_dup_data(WASMModuleInstanceCommon *module_inst,
+    // TODO
+#endif
+}
+#endif
```
把这个文件的改动以及wasm_runtime_common.h的改动删了吧，shared_heap_wrapper.c里面直接调用wasm_runtime_shared_heap_malloc和wasm_runtime_shared_heap_free就可以

yamt committed:
spec_test_on_nuttx.yml: Disable riscv32_ilp32f for now (#3777)

It seems failing too frequently.

cf. https://github.com/bytecodealliance/wasm-micro-runtime/issues/3776

yamt committed:
Appease GCC -Wformat (#3783)

I'm not sure we want to use C99 %tu here.
While C99 %zu is more widely used in WAMR, %tu is rare (if any)
and I'm not sure if it's ubiquitously implemented in platforms
we support.

WenLY1 committed:
fix some issues

Signed-off-by: wenlingyun1 <wenlingyun1@xiaomi.com>

WenLY1 committed:
fix some issues

Signed-off-by: wenlingyun1 <wenlingyun1@xiaomi.com>

wenyongh commented on core/iwasm/interpreter/wasm_runtime.c:
```diff
@@ -3164,6 +3164,9 @@ wasm_deinstantiate(WASMModuleInstance *module_inst, bool is_sub_inst)
         wasm_exec_env_destroy(module_inst->exec_env_singleton);
     }
+#if WASM_ENABLE_SHARED_HEAP != 0
+    wasm_runtime_detach_shared_heap((WASMModuleInstanceCommon *)module_inst);
```
这边也可以不调用，因为detach也只是把module_inst->e->shared_heap指针设置NULL，并没有释放内存等动作。整个module_inst都被free掉，设置它的field也没意义。

wenyongh commented on core/iwasm/interpreter/wasm_loader.c:
```diff
@@ -3233,6 +3233,7 @@ load_memory(const uint8 **p_buf, const uint8 *buf_end, WASMMemory *memory,
 #if WASM_ENABLE_APP_FRAMEWORK == 0
     max_page_count = is_memory64 ? DEFAULT_MEM64_MAX_PAGES : DEFAULT_MAX_PAGES;
 #endif
+
```
不需要的改动，最好改回去

wenyongh commented on core/iwasm/interpreter/wasm_runtime.h:
```diff
@@ -92,6 +92,17 @@ typedef union {
+    void *heap_handle;
+    uint8_t *base_addr;
+    uint32_t size;
+    struct WASMSharedHeap *next;
```
这一句可以作为第一个field放到第一行吗？

loganek committed:
Fix compiler warnings (#3784)

The definition of os_get_invalid_handle in platform_internal.h did not
match the declaration in platform_api_extension.h.

wenyongh commented on core/iwasm/include/wasm_export.h:
```diff
@@ -2110,6 +2119,52 @@ wasm_runtime_detect_native_stack_overflow_size(wasm_exec_env_t exec_env,
+ */
+WASM_RUNTIME_API_EXTERN bool
+wasm_runtime_attach_shared_heap(wasm_module_inst_t module_inst,
+                                void *shared_heap);
```
最好将`void *shared_heap` 改成 `wasm_shared_heap_t shared_heap`
同时内部wasm_runtime_attach_shared_heap_internal, wasm_cluster_attach_shared_heap等也改一下

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,16 +141,370 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+{
+#if WASM_ENABLE_THREAD_MGR != 0
+    wasm_cluster_attach_shared_heap(module_inst, shared_heap);
+    return true;
```
应该是`return wasm_cluster_attach_shared_heap(module_inst, shared_heap);`

WenLY1 replied:
wasm_cluster_attach_shared_heap这个函数内部调用traverse_list遍历所有的shared heap，traverse_list函数是没有返回值的，wasm_cluster_attach_shared_heap这个函数直接返回true可以吗

wenyongh replied:
要不先判断一下当前这个module_inst有没有已经attach了shared heap，有的话return false。否则wasm_cluster_attach_shared_heap并返回true？

WenLY1 replied:
> 要不先判断一下当前这个module_inst有没有已经attach了shared heap，有的话return false。否则wasm_cluster_attach_shared_heap并返回true？

嗯，把这个逻辑放到wasm_cluster_attach_shared_heap函数中了

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,16 +141,370 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+
+bool
+wasm_runtime_attach_shared_heap(WASMModuleInstanceCommon *module_inst,
+                                void *shared_heap)
```
最好把`void *`改成`WASMSharedHeap *`

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,16 +141,370 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+    }
+}
+
+bool
```
最好加static

wenyongh commented on core/iwasm/common/wasm_memory.c:
```diff
@@ -132,16 +141,370 @@ is_bounds_checks_enabled(WASMModuleInstanceCommon *module_inst)
+    return false;
+}
+
+bool
```
最好加static

WenLY1 committed:
fix some issues

Signed-off-by: wenlingyun1 <wenlingyun1@xiaomi.com>

loganek committed:
Implement option for skipping function index in the callstack (#3785)

Also add a script that converts instruction pointers to function indexes (or function names).

https://github.com/bytecodealliance/wasm-micro-runtime/issues/3758

WenLY1 committed:
fix coding-style warning

Signed-off-by: wenlingyun1 <wenlingyun1@xiaomi.com>

no1wudi committed:
Ignore temporary file from aider (#3787)

Aider is AI pair programming in your terminal: https://aider.chat

wenyongh committed:
Merge pull request #3792 from bytecodealliance/main

Merge branch main into dev/shared_heap

WenLY1 committed:
fix "out of bounds memory access"

Signed-off-by: wenlingyun1 <wenlingyun1@xiaomi.com>
